paasta_tools/cli/cmds/mark_for_deployment.py

```diff
@@ -1463,68 +1369,84 @@ def wait_for_deployment(
             ),
             level="event",
         )
-        return
+        return None
 
     print(
         "Waiting for deployment of {} for '{}' to complete...".format(
             git_sha, deploy_group
         )
     )
 
-    deadline = time.time() + timeout
-    if green_light is None:
-        green_light = Event()
-    green_light.set()
-
     with progressbar.ProgressBar(maxval=total_instances) as bar:
-        while time.time() < deadline:
-            _query_clusters(clusters_data, green_light)
-            if not green_light.is_set():
-                raise KeyboardInterrupt
+        instance_done_futures = []
+        with concurrent.futures.ThreadPoolExecutor() as pool:
+            for cluster, instance_configs in instance_configs_per_cluster.items():
+                for instance_config in instance_configs:
+                    instance_done_futures.append(
+                        asyncio.ensure_future(
+                            wait_until_instance_is_done(
+                                pool,
+                                service,
+                                instance_config.get_instance(),
+                                cluster,
+                                git_sha,
+                                instance_config,
+                            ),
+                        )
+                    )
+
+            remaining_instances: Dict[str, List[str]] = {
```

Comment:
Given how the value is being used later on, it looks like it makes more sense for it to be a `set` instead of a `list`.